lib/launcher.ts

```diff
@@ -108,197 +107,170 @@ let initFn = function(configFile: string, additionalConfig: Config) {
+    try {
+      let returned =
+          await helper.runFilenameOrFn_(config.configDir, config.afterLaunch, [exitCode]);
+      if (typeof returned === 'number') {
+        process.exit(returned);
+      } else {
+        process.exit(exitCode);
+      }
+    } catch (err) {
+      logger.error('Error:', err);
+      process.exit(1);
+    }
+  };
 
-        let deferred = q.defer<any>();  // Resolved when all tasks are completed
-        let createNextTaskRunner = () => {
-          let task = scheduler.nextTask();
-          if (task) {
-            let taskRunner = new TaskRunner(configFile, additionalConfig, task, forkProcess);
-            taskRunner.run()
-                .then((result) => {
-                  if (result.exitCode && !result.failedCount) {
-                    logger.error(
-                        'Runner process exited unexpectedly with error code: ' + result.exitCode);
-                  }
-                  taskResults_.add(result);
-                  task.done();
-                  createNextTaskRunner();
-                  // If all tasks are finished
-                  if (scheduler.numTasksOutstanding() === 0) {
-                    deferred.resolve();
-                  }
-                  logger.info(
-                      scheduler.countActiveTasks() + ' instance(s) of WebDriver still running');
-                })
-                .catch((err: Error) => {
-                  logger.error('Error:', (err as any).stack || err.message || err);
-                  cleanUpAndExit(RUNNERS_FAILED_EXIT_CODE);
-                });
-          }
-        };
-        // Start `scheduler.maxConcurrentTasks()` workers for handling tasks in
-        // the beginning. As a worker finishes a task, it will pick up the next
-        // task
-        // from the scheduler's queue until all tasks are gone.
-        for (let i = 0; i < scheduler.maxConcurrentTasks(); ++i) {
-          createNextTaskRunner();
-        }
-        logger.info('Running ' + scheduler.countActiveTasks() + ' instances of WebDriver');
+  let totalTasks = scheduler.numTasksOutstanding();
+  let forkProcess = false;
+  if (totalTasks > 1) {  // Start new processes only if there are >1 tasks.
+    forkProcess = true;
+    if (config.debug) {
+      throw new ConfigError(
+          logger, 'Cannot run in debug mode with multiCapabilities, count > 1, or sharding');
+    }
+  }
 
-        // By now all runners have completed.
-        deferred.promise
-            .then(function() {
-              // Save results if desired
-              if (config.resultJsonOutputFile) {
-                taskResults_.saveResults(config.resultJsonOutputFile);
+  const createNextTaskRunner = async () => {
+    return new Promise(resolve => {
+      let task = scheduler.nextTask();
+      if (task) {
+        let taskRunner = new TaskRunner(configFile, additionalConfig, task, forkProcess);
+        taskRunner.run()
+            .then(async (result) => {
```

Comment:
why you don't want to use `async ... await` here?

Reply:
Good idea. Changing this to async / await